src/httpserver.h

```diff
@@ -27,168 +34,316 @@ static const int DEFAULT_HTTP_WORKQUEUE=64;
- * queue timers or custom events.
- */
-struct event_base* EventBase();
+class HTTPHeaders
+{
+public:
+    std::optional<std::string> Find(const std::string key) const;
+    void Write(const std::string key, const std::string value);
+    void Remove(const std::string key);
+    bool Read(util::LineReader& reader);
+    std::string Stringify() const;
+
+private:
+    std::map<std::string, std::string, util::CaseInsensitiveComparator> m_map;
+};
+
+class HTTPResponse
+{
+public:
+    int m_version_major;
+    int m_version_minor;
+    HTTPStatusCode m_status;
+    std::string m_reason;
+    HTTPHeaders m_headers;
+    std::vector<std::byte> m_body;
+    bool m_keep_alive{false};
+
+    std::string StringifyHeaders() const;
+};
+
+class HTTPClient;
 
-/** In-flight HTTP request.
- * Thin C++ wrapper around evhttp_request.
- */
 class HTTPRequest
 {
-private:
-    struct evhttp_request* req;
-    const util::SignalInterrupt& m_interrupt;
-    bool replySent;
+public:
+    std::string m_method;
+    std::string m_target;
+    // Default protocol version is used by error responses to unreadable requests
+    int m_version_major{1};
+    int m_version_minor{1};
+    HTTPHeaders m_headers;
+    std::string m_body;
+
+    // Keep a pointer to the client that made the request so
+    // we know who to respond to.
+    std::shared_ptr<HTTPClient> m_client;
+    explicit HTTPRequest(std::shared_ptr<HTTPClient> client) : m_client(client) {};
+    // Null client for unit tests
+    explicit HTTPRequest() : m_client(nullptr) {};
+
+    // Readers return false if they need more data from the
+    // socket to parse properly. They throw errors if
+    // the data is invalid.
+    bool LoadControlData(LineReader& reader);
+    bool LoadHeaders(LineReader& reader);
+    bool LoadBody(LineReader& reader);
 
+    // These methods reimplement the API from http_libevent::HTTPRequest
+    // for downstream JSONRPC and REST modules.
+    std::string GetURI() const {return m_target;};
+    CService GetPeer() const;
+    HTTPRequestMethod GetRequestMethod() const;
+    std::optional<std::string> GetQueryParameter(const std::string& key) const;
+    std::pair<bool, std::string> GetHeader(const std::string& hdr) const;
+    std::string ReadBody() const {return m_body;};
+    void WriteHeader(const std::string& hdr, const std::string& value);
+
+    // Response headers may be set in advance before response body is known
+    HTTPHeaders m_response_headers;
+    void WriteReply(HTTPStatusCode status, std::span<const std::byte> reply_body = {});
+    void WriteReply(HTTPStatusCode status, const char* reply_body)
+    {
+        auto reply_body_view = std::string_view(reply_body);
+        std::span<const std::byte> byte_span(reinterpret_cast<const std::byte*>(reply_body_view.data()), reply_body_view.size());
+        WriteReply(status, byte_span);
+    }
+    void WriteReply(HTTPStatusCode status, const std::string& reply_body)
```

Comment:
Can be removed if we'll add `WriteReply(HTTPStatusCode status, std::string_view reply_body_view)` (see above).